flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -157,7 +158,7 @@
 
 	private final FatalErrorHandler fatalErrorHandler;
 
-	private final ClassLoader userCodeLoader;
+	private final UserCodeClassLoader userCodeLoader;
```

Comment:
How is the `UserCodeClassLoader` going to be used in `JobMaster`?
atm I see only `asClassLoader` usages.
is there a further plan how to expose the `UserCodeClassLoader` in JM user hooks?

Reply:
You are right, we could revert these changes and call the `.asClassloader()` method when passing the classloader into the factory.

I'm not aware of plans to use the UserCodeCLassloader in JM hooks. (We could pass it into the InitializeOnMaster.initializeGlobal() interface, but that would break a public interface)

I also could not find a strong argument for passing the `UserCodeClassLoader` into the `JobMaster`. I guess this is a question of taste, rather than strong logical arguments.
The benefit of the current approach is that we have an extra information about the type of ClassLoader we are using.

I'm undecided here. If you believe it is better to revert this changes, I can do it.

Reply:
We could add another default method to `InitializeOnMaster` which falls back to the existing method.
There is also some checkpoint global completion hook on JM side.
Generally, I believe running of user code in JM side does not differ too much from running in TM and can lead to similar metaspace leaks.
wdyt?

Reply:
Yes, user code running on the JobManager can cause exactly the same problems.

But I'm not sure if it makes sense to provide the shutdown hooks at these locations in the JobManager right now: We currently have no use-case for them.
This change here is done for the Kinesis producer, which runs on the TaskManagers only.

I would vote for introducing this new API in the JobManager once we need it (it is a fairly simple change based on this one, but we can keep the complexity lower for now)

Reply:
ok, sounds good to me

Reply:
Reverting the JM changes is a good point (wasn't aware of that). Now I get the bigger picture of your comments :)